Add a CMake build system to ModEM

MiCurry opened the pull request:
This pull request adds a CMake build system to ModEM. Currently it has the ability to build various version of ModEM.

Can build:
* Dimensions
  * 2D - Not tested
  * 3D 
* Forward Flavors
  * MF
  * SP
  * SP2
* GPU Versions
  * CUDA - Built, but not fully tested
  * HIP - Not tested
* Compilers
  * GFortran
  * Intel/Ifort
* Release/Debug
  * Release
  
Can not build
* Debug
* Spherical Version

This build system has some major advantages over the old configure build system:

* CMake is a well established, widely adopted, and well known build system
  * Will make it easier for future developers/users to build and use ModEM quickly
* Built-In Cross Language Support
  * We can now easily compile C, CUDA, HIP or other languages associated with ModEM
  * Configure had no way to add non-fortran languages into the Makefile, requiring manual editing.
* Package finding
  * CMake can easily help users when finding necessary packages (LAPACK, BLAS libs)
* Portability
  * Able to create various build files: Makefiles, Ninja, Visual Studio, etc.

I think the biggest benefit will be the ability for ModEM to easily be compiled with non-Fortran languages, that should help us bring the CUDA and HIP code more into development.

Note this PR should be merged in after #28

MiCurry commented:
I still need to do some additional testing, but so far I'm confident this can build most ModEM versions.

@dong-hao I was able to build a version of ModEM with the CUDA code/interfaces, but I am still not able to run it. If you get the chance, it would be great if you could try and create a CUDA and HIP capable version of ModEM using CMake. No rush or worries if you can't get to it, but would be great if you did.

You can create a Cuda/HIP Makefile by running the following:

```bash
$ # For CUDA
$ cd ModEM
$ mkdir build-cuda
$ cd build-cuda
$ cmake .. -B . -DBUILD_GPU=CUDA
$ ... lots of output
$ make
```

```bash
$ # For HIP
$ cd ModEM
$ mkdir build-hip
$ cd build-hip
$ cmake .. -B . -DBUILD_GPU=HIP
$ .. output...
$ make
```

You may need to specify the compiler you wish to use: `-DCMAKE_Fortran_COMPILER=mpifort`.  If it has trouble finding BLAS/LAPACK you can tell it to look in a directory: `-DMAKE_PREFIX_PATH="/path/to/install/lib”

If you have any errors or issues feel free to paste them here! No worries if it takes a while for you to get to this, figured I'd share it and see if I can get some feedback!

MiCurry commented:
> Thanks @MiCurry for the introduction of CMake! That's a solid step towards a pipelined CI/CD flow for the ModEM. Glad to help on the CUDA/HIP parts - indeed I got my hands full with the paperworks/reports at this time of the year. I will try to get back to it later.
> 
> Cheers, Hao

Thanks @dong-hao! Good luck with the paperwork/reports at the end of the year!

dong-hao commented:
Hi @MiCurry , 

Sorry for the delayed response - I finally found a little time to test the CMake branch - for now I can confirm that: 

1. The code can be configured and built using command: “cmake -B . -DBUILD_GPU=CUDA -DCMAKE_Fortran_COMPILER=mpifort"

2. On the other hand, I believe the macro is not processed in the current CMake list - while I am far from familiar with CMake, I think we need some “add_definitions()” command to tell the preprocessors to deal with the GPU code, no? Also, the libraries are not supplied, either - guess we also need to add a couple of "-l" options using add_link_options()?  

3. Additionally, somehow CMake did not figure out the dependancies of the modules - probably due to the macro problem, as well?  It only compiles the CUDA kernel after the compiling of Mod3DMT.f90 (!) So it should basically be a SP2 CPU code now (seems to work well, by the way): 
	…
	[ 96%] Building Fortran object f90/CMakeFiles/Mod3DMT_SP2.dir/Mod3DMT.f90.o
	**[ 98%] Building CUDA object f90/CMakeFiles/Mod3DMT_SP2.dir/3D_MT/FWD_SP2/kernel_c.cu.o**
	[100%] Linking Fortran executable Mod3DMT_SP2
	[100%] Built target Mod3DMT_SP2

2. This can also be confirmed using ldd, so even if we have built with the GPU kernels, the code should not work as we don't have the libs plugged:  
        ldd ~/bin/ModEM.gpu.cmake 
	linux-vdso.so.1 (0x000074ce94cf3000)
	/lib/$LIB/liblsp.so => /lib/lib/x86_64-linux-gnu/liblsp.so (0x000074ce94800000)
	liblapack.so.3 => /lib/x86_64-linux-gnu/liblapack.so.3 (0x000074ce94000000)
	libmpi_mpifh.so.40 => /lib/x86_64-linux-gnu/libmpi_mpifh.so.40 (0x000074ce94ac1000)
	libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x000074ce93c00000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x000074ce94717000)
	libmvec.so.1 => /lib/x86_64-linux-gnu/libmvec.so.1 (0x000074ce93b07000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x000074ce94a91000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000074ce93800000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x000074ce94a8c000)
	libblas.so.3 => /lib/x86_64-linux-gnu/libblas.so.3 (0x000074ce93f59000)
	libmpi.so.40 => /lib/x86_64-linux-gnu/libmpi.so.40 (0x000074ce936ce000)
	libopen-pal.so.40 => /lib/x86_64-linux-gnu/libopen-pal.so.40 (0x000074ce93a53000)
	/lib64/ld-linux-x86-64.so.2 (0x000074ce94cf5000)
	libopen-rte.so.40 => /lib/x86_64-linux-gnu/libopen-rte.so.40 (0x000074ce93612000)
	libhwloc.so.15 => /lib/x86_64-linux-gnu/libhwloc.so.15 (0x000074ce94a29000)
	libevent_core-2.1.so.7 => /lib/x86_64-linux-gnu/libevent_core-2.1.so.7 (0x000074ce93f24000)
	libevent_pthreads-2.1.so.7 => /lib/x86_64-linux-gnu/libevent_pthreads-2.1.so.7 (0x000074ce94a22000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x000074ce94a06000)
	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x000074ce93a20000)
	libcap.so.2 => /lib/x86_64-linux-gnu/libcap.so.2 (0x000074ce9470a000)

	So unfortunately the cmake does not work yet for CUDA (or HIP). I would recommend adding the macro and check the link libraries, first.

dong-hao commented:
Thanks @MiCurry for the update - I will try to test the CUDA (and maybe HIP) version and report back.

MiCurry commented:
> Thanks @MiCurry for the update - I will try to test the CUDA (and maybe HIP) version and report back.

I just added support for an additional flag to take in a Cuda Compute Capability, I also updated the documentation page with information on how to compile with Cuda.

I took a little bit of time trying to get HIP working on the USGS machine, but it looks like it will be more involved as there aren't any native AMD machines and I'll need to install ROMc myself. I was able to install HIP, but I might save installing ROMc myself for another day.

I'm not sure if the HIP build will work, but let me know if it does @dong-hao!

dong-hao commented:
Thanks @MiCurry 

I can confirm that the CUDA version compiled with configuration:

`cmake -B . -DBUILD_GPU=CUDA -DCMAKE_Fortran_COMPILER=mpifort`

works without any problem on my local machine. Well done! 

As for the HIP/ROCm - It is well known that HIP libraries needs some extra efforts to work with (there's some reason that NVIDIA won most of the market share, I guess). Curiously, for HIP I got an error message of: 

```
[ 20%] Building Fortran object f90/CMakeFiles/Mod3DMT_SP2.dir/UTILS/file_units.f90.o
[ 22%] Building Fortran object f90/CMakeFiles/Mod3DMT_SP2.dir/MPI/Declaration_MPI.f90.o
/home/donghao/src/git/ModEM.cmake/f90/MPI/Declaration_MPI.f90:10:10:
   10 |      use hipFortMap
      |          1
Fatal Error: Cannot open module file ‘hipfortmap.mod‘
```
So I took a look at the CMakelists.txt in 3D/FWD_SP2 and it indeed seemed to miss the C-Fortran translation code for HIP: 

```
if (BUILD_GPU STREQUAL CUDA)
    target_sources(${MODEM_EXE}
        PRIVATE kernel_c.cu
        PRIVATE cudaFortMap.f90
    )
elseif (BUILD_GPU STREQUAL HIP)
    target_sources(${MODEM_EXE}
        PRIVATE kernel_c.hip
    )
endif()
```
Adding a line of `PRIVATE hipFortMap.f90 ` seems to fix the issue and now the HIP version also work without problems. 



> I just added support for an additional flag to take in a Cuda Compute Capability, I also updated the documentation page with information on how to compile with Cuda.
> 
> I took a little bit of time trying to get HIP working on the USGS machine, but it looks like it will be more involved as there aren't any native AMD machines and I'll need to install ROMc myself. I was able to install HIP, but I might save installing ROMc myself for another day.
> 
> I'm not sure if the HIP build will work, but let me know if it does @dong-hao!

tjhei commented:
This is what I get:
```
$ cmake ..
-- The Fortran compiler identification is GNU 13.3.0
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /usr/bin/f95 - skipped
-- Building BUILD_SP2
-- Looking for Fortran sgemm
-- Looking for Fortran sgemm - not found
-- Looking for Fortran sgemm
-- Looking for Fortran sgemm - found
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libblas.so  
-- Looking for Fortran cheev
-- Looking for Fortran cheev - found
-- Found LAPACK: /usr/lib/x86_64-linux-gnu/liblapack.so;/usr/lib/x86_64-linux-gnu/libblas.so  
-- Configuring done (1.0s)
-- Generating done (0.0s)
-- Build files have been written to: /ssd/cig/ModEM/build
$ ninja
[3+59/114 0.247s] Building Fortran object f90/CMakeFiles/Mod3DMT_SP2.dir/UTILS/math_constants.f90.o
/ssd/cig/ModEM/f90/UTILS/math_constants.f90:24:10:

   24 |   ! Earth's surface for this column)
      |          1
Warning: missing terminating ' character
[8+60/114 0.269s] Building Fortran object f90/CMakeFiles/Mod3DMT_SP2.dir/MPI/Declaration_MPI.f90.o
FAILED: f90/CMakeFiles/Mod3DMT_SP2.dir/MPI/Declaration_MPI.f90.o f90/obj/declaration_mpi.mod 
/usr/bin/f95 -I/ssd/cig/ModEM/f90/MPI -Jf90/obj -cpp -O3 -ffree-line-length-none -fallow-argument-mismatch -fpreprocessed -c f90/CMakeFiles/Mod3DMT_SP2.dir/MPI/Declaration_MPI.f90-pp.f90 -o f90/CMakeFiles/Mod3DMT_SP2.dir/MPI/Declaration_MPI.f90.o
/ssd/cig/ModEM/f90/MPI/Declaration_MPI.f90:231:19:

  231 |     ! note we didn't use the MPI-3.0 features (MPI_GET/PUT)
      |                   1
Warning: missing terminating ' character
/ssd/cig/ModEM/f90/MPI/Declaration_MPI.f90:6:10:

    6 |      use mpi
      |          1
Fatal Error: Cannot open module file ‘mpi.mod’ for reading at (1): No such file or directory
compilation terminated.
```

If I turn off MPI, I can build (with hundreds of warnings - they are probably worth addressing/silencing).
If I manually specify ``-DCMAKE_Fortran_COMPILER=mpifort`` I can build with MPI.

The warnings are related to using ninja and should be fixable by using ``.F90`` instead of ``.f90`` files: https://gitlab.kitware.com/cmake/cmake/-/issues/17466#note_438832 

I can confirm that the warnings are gone when using ``make`` instead of ``ninja``. :+1:

bangerth commented:
How long does compiling each of these executables take?

MiCurry commented:
Thanks for the review as well @bangerth! I've update the code to use `target_compile_options`, `target_link_options` and `target_link_libraries` and to use `CMAKE_BUILD_TYPE`.

I'm making a note here, that one notable missing flag for the Intel compiler is `-xHost` or any version `-m` (`-mavx`, `-msse3`, `-msse`, etc. etc.). However, this might be good to put in the documentation, which I will do.
 
The ifort classic compiler libraries unfortunately don't have any nice CMake `find_package` so instead I've used `find_library` with hinting that the ifort libraries should be inside the `MKLROOT` environment variable, which is set in Intel's `setvars.sh` [script](https://www.intel.com/content/www/us/en/docs/oneapi/programming-guide/2023-0/use-the-setvars-script-with-linux-or-macos.html).

> As for the "forward flavor" (which is actually more like "forward formulation" for matrix free or sparse matrix, etc), we can of course try to build all different ones at once. Although most people will just need the SP2, for computational efficiency. It's nice to have an "option" for streamline testings, of course. something like:
> 
> `Set the forward flavor: < MF | SP | SP2 | ALL > `

I've changed `FORWARD_FLAVOR` to be `FORWARD_FOMULATION`, I like that much better. 

I also like the idea of having `ALL` here in the forward flavor. However I think at this time I'd like to leave it as it is. I think this is a good first version. I think I'd prefer to have the CMake build only build one executable, at least by default, for the user's convenience. As @dong-hao most people will want to run the fastest formulation, SP2. I think leaving the build system as simple as possible will help new users get started. Regardless, it will be easy enough to build each forward formulation in automated tested.

Thanks again @tjhei, and @bangerth for the outside review! And thanks @dong-hao for testing it with Cuda and HIP on your machines and reviewing it.

I've pushed one last commit and I think this PR is looking good to go to be merged in. I think we have a few things to work on in future versions but I think this will be a good first version. Feel free to give it one last look over if you wish!

MiCurry commented:
> How long does compiling each of these executables take?

With the Intel compiler and full optimization it looks like they take about ~2 mins.

dong-hao commented on CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+        # Note: The removal of decimal places.
+        #
+        # One option here would be to use 'native', but on HPC systems, there might not be a 
+        # GPU avaliable where one compiles the code as they are only avaliable on compute nodes,
```
avaliable --> available (twice)

MiCurry commented on f90/3D_MT/FWD_SP2/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+    )
+elseif (BUILD_GPU STREQUAL HIP)
+    target_sources(${MODEM_EXE}
+        PRIVATE kernel_c.hip
```
Added!

MiCurry commented:
> Also, it would be nice to add the option to configure the IO options: we have ascii, binary (almost no one uses it, though) and HDF5 for now. Of course, we don't need to fix those in one PR - let's deal with them later.

I think it might be better to have these as runtime options rather than changing them during building. If the user builds without HDF5, then they could have the option to write out in ASCII or Binary and if they do build with HDF5, then they could choose between the HDF5, ASCII and Binary when they run. We could even have an option to switch between the different ASCII formats WSINV3DMT (the current default) and WingLink/Mackie Format (Which requires altering includes in the source code to switch).

We'd of course have to come up with a way for the user to specify this option. Perhaps we could include them in the forward and inversion control files. I have thought maybe switching to Namelists might be nice, but that would redesign the user interface. 

I'll make an issue for this, it would be nice to have a discussion about possible ways to do this when we get around to it.
